src/ngraph/function.cpp

```diff
@@ -81,6 +81,25 @@ void Function::validate_nodes_and_infer_types()
     ngraph::validate_nodes_and_infer_types(get_ops());
 }
 
+Function::~Function()
+{
+    // Nodes are linked in order from Results to Parameters or backwards through the graph.
+    // If you delete the nodes by simply releasing the results then the nodes recursively
+    // traverse the tree freeing each node's inputs. This will exceed a process's stack allocation
+    // for a sufficiently large graph.
+    // This deleter makes a list of nodes which can then be traversed in forward order to
+    // delete nodes. The only memory cost is the size of the list, but this is allocated on the
+    // heap.
+    std::list<Node*> list;
+
+    traverse_nodes(this, [&](shared_ptr<Node> n) { list.push_front(n.get()); });
```

Comment:
I think we should include the control deps here.

Reply:
This only works if the last reference to the nodes is from the function.
Suppose we had a `Node::free_children(set<shared_ptr<Node>> nodes)` that adds each child node (inputs and control deps) to nodes and then releases them from the inputs. Then it goes through nodes. If `use_count` is 1, it calls free_children on the node, and it removes node from nodes no matter what `node_count` was. When nodes is empty, it returns. Make ~Node call free_children on itself with an empty set. I think that will take care of all the delete cases.

Reply:
@diyessi If by `works` you mean actually frees the nodes then you are correct. Since the nodes are shared_ptr we can't actually delete them as they may be used by others. This is by design. What we can do is unlink them so that each node is now standalone with no references to other nodes. At some later time when the last user of each node releases the last shared_ptr then the node will correctly be deleted. The Function does not own the nodes, it shares them with other.